borg/crypto.pyx

```diff
@@ -63,93 +83,244 @@ def num_aes_blocks(int length):
+            raise MemoryError
+        try:
+            if not HMAC_Init_ex(&self.hmac_ctx, self.mac_key, 32, EVP_sha256(), NULL):
+                raise Exception('HMAC_Init_ex failed')
+            if not HMAC_Update(&self.hmac_ctx, aad, alen):
+                raise Exception('HMAC_Update failed')
+            if not HMAC_Update(&self.hmac_ctx, <const unsigned char *> idata.buf+32, ilen-32):
+                raise Exception('HMAC_Update failed')
+            if not HMAC_Final(&self.hmac_ctx, hmac_buf, NULL):
+                raise Exception('HMAC_Final failed')
+            if CRYPTO_memcmp(hmac_buf, idata.buf, 32):
+                raise Exception('Authentication failed')
+            iv = self.fetch_iv(<unsigned char *> idata.buf)
+            if not EVP_DecryptInit_ex(&self.ctx, EVP_aes_256_ctr(), NULL, self.enc_key, iv):
                 raise Exception('EVP_DecryptInit_ex failed')
+            offset = 0
+            rc = EVP_DecryptUpdate(&self.ctx, odata+offset, &olen, <const unsigned char*> idata.buf+40, ilen-40)
+            if not rc:
+                raise Exception('EVP_DecryptUpdate failed')
+            offset += olen
+            rc = EVP_DecryptFinal_ex(&self.ctx, odata+offset, &olen)
+            if rc <= 0:
+                # this error check is very important for modes with padding or
+                # authentication. for them, a failure here means corrupted data.
+                # CTR mode does not use padding nor authentication.
+                raise Exception('EVP_DecryptFinal_ex failed')
+            offset += olen
+            return odata[:offset]
+        finally:
+            free(odata)
+            PyBuffer_Release(&idata)
+
+    cdef fetch_iv(self, unsigned char * idata):
+        # fetch lower 8 bytes of iv and add upper 8 zero bytes
+        return b"\0" * 8 + idata[32:40]
+
+    cdef store_iv(self, unsigned char * odata, unsigned char * iv):
+        # store only lower 8 bytes, upper 8 bytes are assumed to be 0
+        cdef int i
+        for i in range(8):  # XXX easier way?
+            odata[32+i] = iv[8+i]
+
+
+cdef class AES_GCM_256_GMAC:
+    # Layout: GMAC 16 + IV 12 + CT (borg 1.2)
+    # additionally, each chunk starts with a type byte,
+    # which is not passed to this code or added by this code.
+
+    cdef EVP_CIPHER_CTX ctx
+    cdef unsigned char *enc_key
+    cdef unsigned char *iv
+
+    def __cinit__(self, mac_key, enc_key, iv=None):
+        assert mac_key is None
+        assert isinstance(enc_key, bytes) and len(enc_key) == 32
+        assert iv is None or isinstance(iv, bytes) and len(iv) == 12
+        self.enc_key = enc_key
+        if iv is not None:
+            self.iv = iv
+        EVP_CIPHER_CTX_init(&self.ctx)
+
+    def __dealloc__(self):
+        EVP_CIPHER_CTX_cleanup(&self.ctx)
 
     @property
-    def iv(self):
-        return self.ctx.iv[:16]
+    def current_iv(self):
+        raise NotImplemented  # gcm mode does not maintain/increment the counter in self.ctx.iv
```

Comment:
Since there's no interfacing mandating current_iv(): Just remove the method entirely?

Since this is sensitive/crypto code it might be a good idea to explain in a comment why this doesn't make sense with AES-GCM.